### PR TITLE
Try a new layout for compiler flags for OCaml 4.12.0 (and add 4.12.0~alpha1)

### PR DESCRIPTION
**Pull request by AltGr**
This PR is a continuation of #16753 and the general attempt to reduce the number of `ocaml-variants` packages in the repository.

It's proposed to test this layout with the 4.12 development releases of OCaml - if successful, there can be further discussion for updating 4.11 and earlier to use the same idea.

What you can now do:
**opam 2.1**
- `opam switch create alpha-testing 4.12.0~alpha1`to get the normal compiler release.
- `opam switch create alpha-testing-flambda ocaml-variants.4.12.0~alpha1+options ocaml-option-flambda` to get alpha1 with flambda enabled.

**opam 2.0**
- [`opam repo add beta git+https://github.com/ocaml/ocaml-beta-repository.git`]
- `opam switch create alpha-testing 4.12.0~alpha1 --repos=default,beta` to get the normal compiler release (i.e. you need the beta repository enabled, as in previous releases)
- `opam switch create alpha-testing-flambda --packages=ocaml-variants.4.12.0~alpha1+options,ocaml-option-flambda --repos=default,beta` 

How it works:
- `ocaml-base-compiler` is unchanged and, as today, compiles and installs OCaml with default options
- The `ocaml-variants.4.12.0+trunk` package and the new `ocaml-variants.4.12.0~alpha1+options` packages both allow compiler options to be customised by also installing `ocaml-options-` packages:
	+ `ocaml-option-32bit` - 32bit build
	+ `ocaml-option-afl` - enable AFL support
	+ `ocaml-option-bytecode-only` - build bytecode compiler only
	+ `ocaml-option-default-unsafe-string` - enable unsafe-string by default
	+ `ocaml-option-flambda` - enable flambda
	+ `ocaml-option-fp` - enable frame pointers
	+ `ocaml-option-musl` - use musl instead of libc
	+ `ocaml-option-nnp` - enable no-naked-pointers mode
	+ `ocaml-option-no-flat-float-array` - disable flat float arrays in the runtime
	+ `ocaml-option-spacetime` (obviously, for 4.12 this option is not available)
	+ `ocaml-option-static` - no dynamic linking
- In particular, this allows options to be combined: for example installing `ocaml-options-32bit` and `ocaml-options-flambda` enables a 32-bit flambda compiler.
- The package `ocaml-options-vanilla` can be used to make `ocaml-variants.4.12.0~alpha1+options` or `ocaml-variants.4.12.0+trunk` behave like `ocaml-base-compiler`. In particular if your base packages (or opam 2.1 switch invariant) include `ocaml-options-vanilla` then you can be sure that the compiler will not be altered if you install a package which depends on an option (for example, a package which depends on `ocaml-options-flambda` will not cause a switch to recompile).
- The development variants for 4.12 have been removed, but no other variants have been at this stage 
- OCaml 4.12 numbers alpha, beta and rc releases using `~name` instead of `+name`. Support for this was added but intentionally not used in 4.11. `ocaml-config` needs a small update to support this. In order to prevent lots of switches needing to recompile, this is done in `ocaml-config.2` which is only available for 4.12.0 onwards.

### CODE REVIEW

File: packages/ocaml/ocaml.4.12.0/opam [hunk 1]
```diff
@@ -6,9 +6,9 @@ and polls it to initialise specific variables like `ocaml:native-dynlink`"""
 maintainer: "platform@lists.ocaml.org"
 depends: [
   "ocaml-config"
-  "ocaml-base-compiler" {= "4.12.0"} |
-  "ocaml-variants" {>= "4.12.0" & < "4.12.1~"} |
-  "ocaml-system" {>= "4.12.0" & < "4.12.1~"}
+  "ocaml-variants" {>= "4.12.0~" & < "4.12.1~"} |
+  "ocaml-system" {>= "4.12.0" & < "4.12.1~"} |
+  "metaocaml" {>= "4.12.0~" & < "4.12.1~"}
```
[the remark this answers is not in the document]
Comment (AltGr) [reply packages/ocaml/ocaml.4.12.0/opam:11]: Well no need, `ocaml-base-compiler` if needed would just be an alias for `ocaml-variants`+`ocaml-options-vanilla` ?
(indeed `metaocaml` is not needed at the moment since the package doesn't exist, on my other branch I moved it to its own package)

File: packages/ocaml-options-flambda-only/ocaml-options-flambda-only.1/opam [hunk 2]
```diff
@@ -0,0 +1,15 @@
+opam-version: "2.0"
```
[the remark this answers is not in the document]
Comment (AltGr) [reply packages/ocaml-options-flambda-only/ocaml-options-flambda-only.1/opam:1]: Indeed. That also would avoid the unfortunate clash with `bytecode-only`

File: packages/ocaml-variants/ocaml-variants.4.12.0+trunk/opam [hunk 3]
```diff
@@ -9,14 +13,36 @@ depends: [
   "ocaml-beta"
```
[the remark this answers is not in the document]
Comment (AltGr) [reply packages/ocaml-variants/ocaml-variants.4.12.0+trunk/opam:13]: You mean `< 2.1` ? In this case, yes!

[the remark this answers is not in the document]
Comment (dra27) [reply packages/ocaml-variants/ocaml-variants.4.12.0+trunk/opam:13]: Yes, oops! 😣

File: packages/ocaml/ocaml.4.12.0/opam (hunk 1 above)
[the remark this answers is not in the document]
Comment (dra27) [reply packages/ocaml/ocaml.4.12.0/opam:11]: For now, I think it's better to keep the existing layout - for example, it wouldn't surprise me to find scripts which behave dependent on whether there's ocaml-base-compiler or ocaml-variants, for example?

[the remark this answers is not in the document]
Comment (AltGr) [reply packages/ocaml/ocaml.4.12.0/opam:11]: So you mean, duplicating the (actual) compiler definition into an `ocaml-base-compiler` package ?

**Comment by AltGr** [comment]
Fixed all the trivial issues; not sure yet what to do with `ocaml-base-compiler`

The biggest use I think could be through the variables defined by the `ocaml` wrapper package (using scripts in `ocaml-config`), i.e. `ocaml:compiler`. It's defined at the moment as:
```
compiler: "%{ocaml-system:installed?system:}%%{ocaml-base-compiler:version}%%{ocaml-variants:version}%"
```
of course, this should be fixed if `ocaml-base-compiler` can be installed together with `ocaml-variants` (although if the former is not in the dependencies, the variable should remain undefined here).
Rebuilding the same string could be done artificially through stg like
```
"%{ocaml-variants:version}%%{ocaml-options-foo:installed?+foo:}%%{ocaml-options-bar:installed?+bar:}%..."
```

Another question: do you think we need `only-*` packages for all the options ??

File: packages/ocaml/ocaml.4.12.0/opam (hunk 1 above)
[the remark this answers is not in the document]
Comment (dra27) [reply packages/ocaml/ocaml.4.12.0/opam:11]: Sort of - but it wouldn't need any of the `configure` switches and presumably should conflict all the `options` packages?

**Comment by dra27** [comment]
For now, I think we should keep `ocaml-base-compiler` only to keep the change strictly limited to `ocaml-variants` - I completely agree that it becomes less or even un- necessary.

For the `only` packages, once you have one, I'm not sure I can justify why you don't have the others! Which one's would you exclude from wanting to combine? 🙂

**Comment by AltGr** [comment]
I agree on keeping `ocaml-variants`... I was only wondering if it was better to have it as a duplicate definition (more conservative, since it means this path is not changed at all from before), or as a virtual package (forcing `ocaml-options-vanilla`), in which case it would still be accessible to the user, but move outside of the deps of `ocaml`.

About `only` packages, I was assuming that they were only useful for potential "release" or "production" switches where you want to make certain your compiler doesn't get polluted by unwanted options. I wouldn't see the point for `afl`, for example (but yes, e.g.  `static` would make perfect sense here). Dunno, just trying to not multiplicate packages too much at once ;)

**Comment by Octachron** [comment]
Just to be sure, the shortest command line to install the vanilla compiler under the new layout will be
```bash
opam switch create alpha-testing --packages=ocaml-base-compiler.4.12.0~alpha1 --repositories=default,beta
```
?

**Comment by AltGr** [comment]
@Octachron you can do shorter :)
- `--packages` shouldn't be needed
- adding the `beta` repo should no longer be needed with 2.1

So it could be reduced to
```
opam switch create alpha-testing ocaml-base-compiler.4.12.0~alpha1
```

To go the "newer way", that could be instead:
```
opam switch create alpha-testing 4.12.0~alpha1 ocaml-options-vanilla
```
(1 - you will need `--packages` in this case for opam 2.0 IIRC; 2 - you can actually skip `options-vanilla`, it is only useful on 2.1 if you want to make certain no compiler flags can be set later on in this switch)

@dra27 one issue I could see is that if we keep `ocaml-variants.4.12.0` and `ocaml-base-compiler.4.12.0`, for the tweaky underspecified command `opam sw cr 4.12.0`, the automatic invariant will become a disjunction of the two and the solver might choose `base-compiler` at the moment of installation. I think this will not _really_ be an issue though, since, due to the conflicts, if you then require an option the solver will safely switch to `ocaml-variants`, but worth noting.)

**Comment by dra27** [comment]
For the announcement, I suggest we stick with the opam-2.0 compatible version - it should be the case by the time OCaml 4.13 is branched that 2.1 is out, so we could for that announcement have the shorter 2.1 syntax and then have the 2.0-compatible version as a footnote, but at this stage I think it's more confusing!

**Comment by dra27** [comment]
For the `ocaml-options-only` packages, the problem is that as soon as someone wants one that isn't there, it's a nuisance! I agree that this PR can readily just have the `flambda` one, to correspond with upgrading the existing variants packages, but I think we should shortly after have it such `ocaml-options-only-bar-foo-widget` definitely exists for any combination of the options (in alphabetical order)

**Comment by dra27** [comment]
This won't matter until release time, but we also need to remember that `ocaml-system` must depend on `ocaml-options-vanilla`

**Comment by AltGr** [comment]
> This won't matter until release time, but we also need to remember that `ocaml-system` must depend on `ocaml-options-vanilla`

Not strictly required, as since all the options have a post-depend on the proper `ocaml-variants`, we still get the guarantee that they can't be used together with `ocaml-system`. But yes, that may make things clearer.

File: packages/ocaml-variants/ocaml-variants.4.12.0~alpha1+options/opam [hunk 4]
```diff
@@ -0,0 +1,77 @@
+opam-version: "2.0"
+synopsis: "Official OCaml release 4.11.1"
```
Comment (Octachron) [inline packages/ocaml-variants/ocaml-variants.4.12.0~alpha1+options/opam:2]: The synopsis should be:

> First alpha release of OCaml 4.12.0

Comment (AltGr) [reply packages/ocaml-variants/ocaml-variants.4.12.0~alpha1+options/opam:2]: Ooops indeed!

**Comment by Octachron** [comment]
The ocaml-config also needs to be updated to make gen_ocaml_config.ml.in aware of ~, otherwise the installation fails with a version mismatch.

File: packages/ocaml-variants/ocaml-variants.4.12.0~alpha1+options/files/ocaml-compiler.install [hunk 5]
```diff
@@ -0,0 +1 @@
+share_root: ["config.cache" {"ocaml/config.cache"}]
```
Comment (dra27) [inline packages/ocaml-variants/ocaml-variants.4.12.0~alpha1+options/files/ocaml-compiler.install:1]: This file needs renaming to `ocaml-variants.install`

File: packages/ocaml-variants/ocaml-variants.4.12.0~alpha1+options/opam [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+  ]
+  [make "-j%{jobs}%" {os != "cygwin"} "world"]
+  [make "-j%{jobs}%" {os != "cygwin"} "world.opt"]
+]
+install: [make "install"]
+url {
+  src: "https://github.com/ocaml/ocaml/archive/4.12.0-alpha1.tar.gz"
+  checksum: "sha256=bee59cb94067410d02f0bc4e7e47e3e878689aabf61e6d2f0cb4316f8563e55d"
+}
+extra-files: ["ocaml-compiler.install" "md5=3e969b841df1f51ca448e6e6295cb451"]
```
Comment (dra27) [inline packages/ocaml-variants/ocaml-variants.4.12.0~alpha1+options/opam:55]:
```suggestion
extra-files: ["ocaml-variants.install" "md5=3e969b841df1f51ca448e6e6295cb451"]
```

**Review by dra27 — COMMENTED** [review]
Spotted a stray `ocaml-compiler`

**Comment by Octachron** [comment]
Another point, installing compiler with
```ocaml
opam switch create 4.12.0~alpha1+flambda --packages=ocaml-variants.4.12.0~alpha1+options,ocaml-options-flambda --repo=beta
```
yields a warning with opam 2.07

> Packages { ocaml-options-flambda.1 } don't have the 'compiler' flag set. Are you sure you want to set them as the compiler base for this switch? [y/N]

Should the option packages set this compiler flag for compatibility's sake?

Also, I have a patch for ocaml-config, would it help if I sent it as a subpr?

**Comment by dra27** [comment]
Oops, sorry - I'd had a discussion with @AltGr offline and not reported the relevant bits back here. I agree that `ocaml-options-*` packages need the `compiler` flag to get rid of that message. That flag means they won't warn about being installed in the base (and won't warn if not installed in the base) of the switch.

I'd also proposed https://github.com/dra27/opam-repository/commit/7bf0c44c70fc140aa719e1f2585991cf7d2806b6 which adds `ocaml-config.2` (is the change similar to yours, I hope?!) and also https://github.com/dra27/opam-repository/commit/529b7e78ca6e7b10e3ba2bf2e7bdf55992d3867b which adds `ocaml-base-compiler.4.12.0~alpha1`

**Comment by Octachron** [comment]
Yep, my changes in ocaml-config were equivalent. Are there any blockers left for a release in the beginning of next week?

**Comment by dra27** [comment]
I don't believe so, no - just needs @AltGr to push the changes here, if you're happy for this PR to turn into the "OCaml 4.12.0 Alpha1" PR!

**Comment by Octachron** [comment]
I think it is better to have an early release for this first alpha, and fix any remaining issues in the following alphas.

**Comment by Octachron** [comment]
After some local tests, the new packages work great.
We might want to add a configuration option for the non-naked-pointer-checker configuration option, but that can be done later.

**Review by kit-ty-kate — CHANGES_REQUESTED** [review]
Several things:
* [ ] The 4.13.0 package need updating too otherwise they'll break with the new changes in ocaml-config
* [ ] I don't think I like the move of the alpha packages from `ocaml-variants` to `ocaml-base-compiler`. I think it should stay stable releases only. Also it's not really consistent with having the `+trunk` package in `ocaml-variants` in my opinion.
* [x] I think `ocaml-option-*` (singular) would fit better for the useflag packages than its plural counterpart
* [x] FreeBSD was re-added to `"CC=cc" {os = "openbsd" | os = "macos"}`. If I remember correctly it was changed after a comment from @hannesm saying freebsd should not require that.
* [ ] For the `ocaml-variants.4.12.0~alpha1+options` package I think it would make more sense if the options were a disjunctions rather than depopts. It would be rather surprising to have a `+options` setup but no actual options.
* [x] The vanilla packages should depend on the new `ocaml-option-vanilla` so that packages can request this and get the correct behaviour.
* [ ] The main issue brought up in https://github.com/ocaml/opam-repository/pull/16753 is still there: the `option-*` package style does not offer any guaranties that the set requested when creating the switch will stay the same. `options` could be change without user's consent by just one package requesting an option and next thing you know your pretty `4.12+flambda` switch does not contain flambda anymore. To fix this I think we should make all the options packages incompatible with the other ones and add new one combining some of them.
* [ ] Do we have a migration plan for opam 2.2 to switch this configuration to a [USE-flag-like](https://wiki.gentoo.org/wiki/Handbook:AMD64/Working/USE) system? I think if the migration will includes a total rearrangement of the compiler packages it should be safe to merge this PR, however if it does not include it I think we should think twice about how it would work with the new features. I'm afraid this is going to bring more backward compatibility issues if this is not thought of before.

**Comment by AltGr** [comment]
Just answering the few points where I have clear answers on the top of my head:

> For the ocaml-variants.4.12.0~alpha1+options package I think it would make more sense if the options were a disjunctions rather than depopts. It would be rather surprising to have a +options setup but no actual options.

I don't see why it should be surprising; the package has been named `+options` at the moment to avoid confusion, but is intended to become the "main" compiler.

> options could be change without user's consent by just one package

Well, I know it's a lost fight; but opam never installs anything without user consent. (Yes, `--yes` is a blank check that amounts to user consent). :smiling_imp: 

> your pretty 4.12+flambda switch does not contain flambda anymore

Options could be added, but not removed if they were specified at switch creation

> To fix this I think we should make all the options packages incompatible with the other ones and add new one combining some of them.

There is `ocaml-options-only-flambda` for this specific purpose; we intend to add similar ones for all options where it makes sense, and some common combinations (also, the current variants could remain for compat, and do something similar). We decided to leave them out at the moment to avoid too much clutter in the PR.

> The vanilla packages should depend on the new ocaml-option-vanilla so that packages can request this and get the correct behaviour.

You mean the older `ocaml-base-compiler` packages ?
I would rather add this as a post-dependency to `ocaml-options-vanilla`, but yes it makes sense (the actual options handle it this way); and actually, the recent addition of `flags: compiler` is broken without it!

**Comment by hannesm** [comment]
> FreeBSD was re-added to `"CC=cc" {os = "openbsd" | os = "macos"}`. If I remember correctly it was changed after a comment from @hannesm saying FreeBSD should not require that

Indeed, the OCaml configuration script nicely detects `cc` on FreeBSD (since 4.10.1), and the `CC=cc` shouldn't be needed (actually neither on macos nor OpenBSD, but someone with such a system should test this before removing it). See https://github.com/ocaml/ocaml/pull/9437 and https://github.com/ocaml/opam-repository/pull/16575 for details if interested.

**Comment by dra27** [comment]
Thanks, @kit-ty-kate! This PR is supposed to be an incremental step towards an ultimate neat solution so the intention with some of the corners has been to be paranoid about preserving existing use-cases. Thanks to @Octachron's agreement, the idea is to experiment with solely within the non-critical 4.12/4.13 development release packages, so from the OCaml-side we are willing to take a risk and if it turns out it's disastrous for some unforeseen reason then revert later as part of the alpha/beta/rc cycle. 

> * [ ]  The 4.13.0 package need updating too otherwise they'll break with the new changes in ocaml-config

There shouldn't be any breakage - what are you seeing? `ocaml-config.2` merely adds the ability to detect `~` in the version string. The only difference here is that switches with `ocaml-variants.4.13.0+trunk` will ask to recompile if they're unlocked, but I view that as low-priority for the OCaml "dev-repo" package. 

> * [ ]  I don't think I like the move of the alpha packages from `ocaml-variants` to `ocaml-base-compiler`. I think it should stay stable releases only. Also it's not really consistent with having the `+trunk` package in `ocaml-variants` in my opinion.

I could do with a technical reason. The reason for correcting the semver (i.e. using `~alpha1` instead of `+alpha1`) is precisely to allow upgrades - that is, a switch on `ocaml-base-compiler.4.12.0~alpha1` can naturally upgrade to `alpha2` and on to release. As with previous versions, you can only end up with an alpha/beta/rc release if you explicitly request it (either by allowing `ocaml-beta.enabled` to be installed for opam 2.0 or by explicitly requesting `4.12.0~alpha1` as the version).

The inconsistency with `+trunk` is also an artefact of backwards compatibility. The `+trunk` packages themselves are an artefact of OPAM 1.x and the intention is to use `--dev-repo` (they're moving targets, so depending on `ocaml-variants.4.13.0+trunk` is largely meaningless) but that's an orthogonal change to this.

> * [ ]  I think `ocaml-option-*` (singular) would fit better for the useflag packages than its plural counterpart

I like this idea!

> * [ ]  For the `ocaml-variants.4.12.0~alpha1+options` package I think it would make more sense if the options were a disjunctions rather than depopts. It would be rather surprising to have a `+options` setup but no actual options.

@AltGr has also commented on this. The ultimate aim if we go with this experiment is that there would no longer be `ocaml-base-compiler` and `ocaml-variants+options`. The only reason for having the `+options` is to maximise the backwards compatibility which means keeping `opam switch create 4.12.0~alpha1` consistent. We can potentially bike-shed the exact name for `+options`!

> * [x]  The vanilla packages should depend on the new `ocaml-option-vanilla` so that packages can request this and get the correct behaviour.

I don't completely follow - do you mean that `ocaml-base-compiler` et al should depend on it? In particular, my impression is that packages should be discouraged (or even prohibited) from depending on `ocaml-option-vanilla` - what package would you have in mind that has the "right" to say "I must be installed with a vanilla compiler" (for example, ruling out installation on 32bit or in a bytecode-only switch). Rather, such packages should explicitly _conflict_ the option they cannot be built with (i.e. "I depend on `ocaml-option-no-flat-float-array` because I don't work without it" or "I conflict `ocaml-option-nnp` because I definitely use naked pointers")

> * [ ]  The main issue brought up in #16753 is still there: the `option-*` package style does not offer any guaranties that the set requested when creating the switch will stay the same. `options` could be change without user's consent by just one package requesting an option and next thing you know your pretty `4.12+flambda` switch does not contain flambda anymore. To fix this I think we should make all the options packages incompatible with the other ones and add new one combining some of them.

There are `ocaml-options-only` packages to deal (I think) with this situation - more can be added if they're requested (indeed, if really wanted, they can all be generated with a script!).

> * [ ]  Do we have a migration plan for opam 2.2 to switch this configuration to a [USE-flag-like](https://wiki.gentoo.org/wiki/Handbook:AMD64/Working/USE) system? I think if the migration will includes a total rearrangement of the compiler packages it should be safe to merge this PR, however if it does not include it I think we should think twice about how it would work with the new features. I'm afraid this is going to bring more backward compatibility issues if this is not thought of before.

There's no need to block a test in pre-release versions of the package, though. I'm not particularly convinced that concern for a future opam feature needs to block changes to an opam 2.0 repository. Whatever feature is brought in, opam 2.0 is in LTS for longer than that, so either the new feature needs a compatibility story with the existing layout or it needs to be trivially possible to maintain both. The point with this layout - ignoring the `hidden-version` flag, is that it utilises existing opam 2.0 facilities.

**Comment by AltGr** [comment]
Note: @camelus rejects the "new broken package: `ocaml-option-spacetime.1`", but that's intended, the option is provisioned for earlier compilers and this PR only adds 4.12, which doesn't support it.

File: packages/ocaml-variants/ocaml-variants.4.12.0+trunk/opam [hunk 7]
```diff
@@ -1,22 +1,48 @@
 opam-version: "2.0"
 synopsis: "latest 4.12 development"
 maintainer: "platform@lists.ocaml.org"
+authors: "Xavier Leroy and many contributors"
```
Comment (dra27) [inline packages/ocaml-variants/ocaml-variants.4.12.0+trunk/opam:4]:
```suggestion
authors: ["Xavier Leroy" "Damien Doligez" "Alain Frisch" "Jacques Garrigue" "Didier Rémy" "Jérôme Vouillon"]
```
(as before)

File: packages/ocaml-variants/ocaml-variants.4.12.0+trunk/opam [hunk 8]
```diff
@@ -1,22 +1,48 @@
...
+homepage: "https://ocaml.org"
+bug-reports: "https://github.com/ocaml/ocaml/issues"
+dev-repo: "git://github.com/ocaml/ocaml"
```
Comment (dra27) [inline packages/ocaml-variants/ocaml-variants.4.12.0+trunk/opam:7]:
```suggestion
dev-repo: "git://github.com/ocaml/ocaml#4.12"
```

File: packages/ocaml-variants/ocaml-variants.4.12.0~alpha1+options/opam [hunk 9]
```diff
@@ -0,0 +1,77 @@
+opam-version: "2.0"
+synopsis: "First alpha release of OCaml 4.12.0"
+maintainer: "platform@lists.ocaml.org"
+authors: "Xavier Leroy and many contributors"
```
Comment (dra27) [inline packages/ocaml-variants/ocaml-variants.4.12.0~alpha1+options/opam:4]:
```suggestion
authors: ["Xavier Leroy" "Damien Doligez" "Alain Frisch" "Jacques Garrigue" "Didier Rémy" "Jérôme Vouillon"]
```

File: packages/ocaml-variants/ocaml-variants.4.12.0~alpha1+options/opam [hunk 10]
```diff
@@ -0,0 +1,77 @@
...
+homepage: "https://ocaml.org"
+bug-reports: "https://github.com/ocaml/ocaml/issues"
+dev-repo: "git://github.com/ocaml/ocaml"
```
Comment (dra27) [inline packages/ocaml-variants/ocaml-variants.4.12.0~alpha1+options/opam:7]:
```suggestion
dev-repo: "git://github.com/ocaml/ocaml#4.12"
```

File: packages/ocaml-base-compiler/ocaml-base-compiler.4.12.0~alpha1/opam [hunk 11]
```diff
@@ -1,36 +1,41 @@
...
   [
     "./configure"
     "--prefix=%{prefix}%"
-    "CC=cc" {os = "openbsd" | os = "macos"}
-    "ASPP=cc -c" {os = "openbsd" | os = "macos"}
-    "--with-afl"
-    "--disable-warn-error"
+    "-C"
+    "CC=cc" {os = "openbsd" | os = "freebsd" | os = "macos"}
+    "ASPP=cc -c" {os = "openbsd" | os = "freebsd" | os = "macos"}
```
Comment (dra27) [inline packages/ocaml-base-compiler/ocaml-base-compiler.4.12.0~alpha1/opam:24]:
```suggestion
    "CC=cc" {os = "openbsd" | os = "macos"}
    "ASPP=cc -c" {os = "openbsd" | os = "macos"}
```

File: packages/ocaml/ocaml.4.12.0/opam [hunk 12]
```diff
@@ -6,9 +6,10 @@ and polls it to initialise specific variables like `ocaml:native-dynlink`"""
 maintainer: "platform@lists.ocaml.org"
 depends: [
   "ocaml-config"
-  "ocaml-base-compiler" {= "4.12.0"} |
-  "ocaml-variants" {>= "4.12.0" & < "4.12.1~"} |
-  "ocaml-system" {>= "4.12.0" & < "4.12.1~"}
+  "ocaml-base-compiler" {>= "4.12.0~" & < "4.12.1~"} |
+  "ocaml-variants" {>= "4.12.0~" & < "4.12.1~"} |
+  "ocaml-system" {>= "4.12.0" & < "4.12.1~"} |
+  "metaocaml" {>= "4.12.0~" & < "4.12.1~"}
```
Comment (dra27) [inline packages/ocaml/ocaml.4.12.0/opam:12]: I think the `metaocaml` line go in a separate PR as and when that's moved

**Review by dra27 — COMMENTED** [review]
Another pass

File: packages/ocaml-base-compiler/ocaml-base-compiler.4.12.0~alpha1/opam (hunk 11 above)
Comment (AltGr) [reply packages/ocaml-base-compiler/ocaml-base-compiler.4.12.0~alpha1/opam:24]: what this re-appeared again ? :fearful:

File: packages/ocaml-config/ocaml-config.2/files/gen_ocaml_config.ml.in [hunk 13]
```diff
@@ -0,0 +1,56 @@
...
+  p "  native: %%b"
+    (Sys.file_exists (ocaml^"opt"^suffix));
+  p "  native-tools: %%b"
+    (Sys.file_exists (ocamlc^".opt"^suffix));
+  p "  native-dynlink: %%b"
+    (Sys.file_exists (Filename.concat libdir "dynlink.cmxa"));
+  p "  stubsdir: %%S"
+    stubsdir;
+  p "  preinstalled: %{ocaml-system:installed}%";
+  p "  compiler: \"%{ocaml-system:installed?system:}%%{ocaml-base-compiler:version}%%{ocaml-variants:version}%%{ocaml-option-32bit:installed?+32bit:}%%{ocaml-option-afl:installed?+afl:}%%{ocaml-option-bytecode-only:installed?+bytecode-only:}%%{ocaml-option-default-unsafe-string:installed?+default-unsafe-string:}%%{ocaml-option-flambda:installed?+flambda:}%%{ocaml-option-fp:installed?+fp:}%%{ocaml-option-musl:installed?+musl:}%%{ocaml-option-nnp:installed?+nnp:}%%{ocaml-option-no-flat-float-array:installed?+no-flat-float-array:}%%{ocaml-option-spacetime:installed?+spacetime:}%%{ocaml-option-static:installed?+static:}%\"";
```
[the remark this answers is not in the document]
Comment (AltGr) [reply packages/ocaml-config/ocaml-config.2/files/gen_ocaml_config.ml.in:54]: hm indeed... :/

[the remark this answers is not in the document]
Comment (AltGr) [reply packages/ocaml-config/ocaml-config.2/files/gen_ocaml_config.ml.in:54]:
```suggestion
  p "  compiler: \"%{ocaml-system:installed?system:}%%{ocaml-base-compiler:version}%%{ocaml-variants:version}%%{ocaml-option-32bit:installed?+32bit:}%%{ocaml-option-afl:installed?+afl:}%%{ocaml-option-bytecode-only:installed?+bytecode-only:}%%{ocaml-option-default-unsafe-string:installed?+default-unsafe-string:}%%{ocaml-option-fp:installed?+fp:}%%{ocaml-option-flambda:installed?+flambda:}%%{ocaml-option-musl:installed?+musl:}%%{ocaml-option-nnp:installed?+nnp:}%%{ocaml-option-no-flat-float-array:installed?+no-flat-float-array:}%%{ocaml-option-spacetime:installed?+spacetime:}%%{ocaml-option-static:installed?+static:}%\"";
```

File: packages/ocaml-option-spacetime/ocaml-option-spacetime.1/opam [hunk 14]
```diff
@@ -0,0 +1,8 @@
+opam-version: "2.0"
+synopsis: "Set OCaml to be compiled with spacetime activated"
+depends: [
+  # Not available until options are ported to earlier compilers
+  "ocaml-variants" {post & >= "4.12.0~" & < "4.12"}
```
[the remark this answers is not in the document]
Comment (AltGr) [reply packages/ocaml-option-spacetime/ocaml-option-spacetime.1/opam:5]: The `>= "4.12.0~"` is here for all `option` packages at the moment, since earlier `ocaml-variants` packages don't support them.
This prevents you from installing `ocaml-variants.4.11.1+*` together with this package and expecting it do to anything.

File: packages/ocaml/ocaml.4.12.0/opam (hunk 12 above)
Comment (AltGr) [reply packages/ocaml/ocaml.4.12.0/opam:12]: Right.
```suggestion
```

Comment (AltGr) [inline packages/ocaml/ocaml.4.12.0/opam:12]:
```suggestion
  "ocaml-system" {>= "4.12.0" & < "4.12.1~"}
```
Right, let's keep `metaocaml` for later.

File: packages/ocaml-option-spacetime/ocaml-option-spacetime.1/opam (hunk 14 above)
[the remark this answers is not in the document]
Comment (dra27) [reply packages/ocaml-option-spacetime/ocaml-option-spacetime.1/opam:5]: Ah, I see!

File: packages/ocaml-base-compiler/ocaml-base-compiler.4.12.0~alpha1/opam [hunk 15]
```diff
@@ -1,36 +1,41 @@
...
+bug-reports: "https://github.com/ocaml/ocaml/issues"
+dev-repo: "git://github.com/ocaml/ocaml#4.12"
 depends: [
   "ocaml" {= "4.12.0" & post}
   "base-unix" {post}
   "base-bigarray" {post}
   "base-threads" {post}
-  "ocaml-beta"
+  "ocaml-beta" {opam-version < "2.1"}
 ]
```
Comment (kit-ty-kate) [inline packages/ocaml-base-compiler/ocaml-base-compiler.4.12.0~alpha1/opam:14]:
```suggestion
  "ocaml-option-vanilla"
]
```

**Review by kit-ty-kate — APPROVED** [review]
Apart from this one small suggestion, after talking about this change more globally I think this experiment looks good to merge

File: packages/ocaml-base-compiler/ocaml-base-compiler.4.12.0~alpha1/opam (hunk 15 above)
Comment (dra27) [reply packages/ocaml-base-compiler/ocaml-base-compiler.4.12.0~alpha1/opam:14]: Note it's remained `ocaml-options-vanilla` (in plural) - so you depend on "vanilla options" (i.e. no options) rather than "the vanilla option" (which, alas, OCaml does not in fact have 🍦)

I think this is already achieved, though, because of the conflict-class, which prevents ocaml-base-compiler switching to ocaml-variants already, @AltGr?

Comment (kit-ty-kate) [reply packages/ocaml-base-compiler/ocaml-base-compiler.4.12.0~alpha1/opam:14]: Oh, do we want to move the remaining `ocaml-options-*` to their singular counterparts before that? I think it's a bit confusing and potential source of errors to have both.

Apart from that, the request itself isn't a blocker, it's more for the sake of completeness. I don't care about it that much.

Comment (AltGr) [reply packages/ocaml-base-compiler/ocaml-base-compiler.4.12.0~alpha1/opam:14]: Ah yes, indeed, in the case of `ocaml-base-compiler`, we have two options:
- Have `ocaml-base-compiler` be just an alias for `ocaml-variants.x+options` & `ocaml-options-vanilla`
- Have it stand-alone as before
- (indeed 3: replace `base-compiler` by the version with options, not needing the `variants.x+options` at all. Not right away though)

The two should be equivalent in what you will get in the resulting switch ; my first take was option 1 here, to avoid duplication of the build instruction, but after some thought we decided to be more conservative and keep the stand-alone definition: it's less disruptive to current users of `ocaml-base-compiler`.

@dra27 yes, `ocaml-variants` and `ocaml-base-compiler` are mutually exclusive as before, and all options require `variants` so this is not needed; I guess @kit-ty-kate  wanted the presence of `options-vanilla` to be a reliable marker that you are on the vanilla case; since the `ocaml-options-vanilla` package has already been made compatible with `ocaml-base-compiler`, there would be no harm in adding this dependency here to that purpose I guess.

File: packages/ocaml-config/ocaml-config.2/opam [hunk 16]
```diff
@@ -0,0 +1,21 @@
...
+homepage: "https://opam.ocaml.org/"
+bug-reports: "https://github.com/ocaml/opam/issues"
+depends: [
+  "ocaml-base-compiler" {>= "4.12.0~"} |
+  "ocaml-variants" {>= "4.12.0~"} |
+  "ocaml-system" {>= "4.12.0~"}
+]
+substs: "gen_ocaml_config.ml"
+extra-files: [
+  ["gen_ocaml_config.ml.in" "md5=07afb6231bbcf399f7f7872e25eb23a1"]
```
Comment (Octachron) [inline packages/ocaml-config/ocaml-config.2/opam:19]: This is the wrong md5 hash, the right one is a4b41e3236593d8271295b84b0969172 .

**Comment by kit-ty-kate** [comment]
Looks alright. Feel free to revert at any point if something is really broken.

Thanks!
